src/khepri_fun.erl

```diff
@@ -330,55 +341,47 @@ add_comment_and_retry(
     compile(Asm1).
 
 add_comment_to_function(
-  Asm, Error,
   [#function{name = Name, arity = Arity, code = Code} = Function | Rest],
-  Name, Arity, Comment, Result) ->
-    Code1 = add_comment_to_code(Asm, Error, Code, Comment, []),
+  {_GeneratedModuleName, Name, Arity},
+  Location, Comment, Result) ->
+    Code1 = add_comment_to_code(Code, Location, Comment),
+    Function1 = Function#function{code = Code1},
+    lists:reverse(Result) ++ [Function1 | Rest];
+add_comment_to_function(
+  [#function{entry = EntryLabel, code = Code} = Function | Rest],
+  EntryLabel, Location, Comment, Result) ->
+    Code1 = add_comment_to_code(Code, Location, Comment),
     Function1 = Function#function{code = Code1},
     lists:reverse(Result) ++ [Function1 | Rest];
 add_comment_to_function(
-  Asm, Error,
-  [Function | Rest], Name, Arity, Comment, Result) ->
+  [Function | Rest], FailingFun, Location, Comment, Result) ->
     add_comment_to_function(
-      Asm, Error, Rest, Name, Arity, Comment, [Function | Result]).
+      Rest, FailingFun, Location, Comment, [Function | Result]).
+
+add_comment_to_code(Code, Location, Comment) ->
+    add_comment_to_code(Code, Location, Comment, []).
 
 add_comment_to_code(
-  Asm, Error,
-  [{label, _} = Instruction | Rest],
-  Comment, Result) ->
-    add_comment_to_code(Asm, Error, Rest, Comment, [Instruction | Result]);
-add_comment_to_code(
-  Asm, Error,
-  [{func_info, _, _, _} = Instruction | Rest],
-  Comment, Result) ->
-    add_comment_to_code(Asm, Error, Rest, Comment, [Instruction | Result]);
-add_comment_to_code(
-  Asm, Error,
-  [{'%', _} = Instruction | Rest],
-  Comment, Result) ->
-    case are_comments_conflicting(Instruction, Comment) of
-        false ->
-            add_comment_to_code(
-              Asm, Error, Rest, Comment, [Instruction | Result]);
-        true ->
-            throw(
-              {conflicting_assembly_annotations,
-               Instruction, Comment, Error, Asm})
-    end;
+  [Instruction | Rest], {Placement, Instruction}, Comment, Result) ->
+    MaybeExistingComment = case Placement of
+                               before -> hd(Result);
+                               'after' -> hd(Rest)
+                           end,
+    Comment1 = merge_comments(Comment, MaybeExistingComment),
+    Rest1 = case Placement of
+              before -> [Comment1, Instruction | Rest];
+              'after' -> [Instruction, Comment1 | Rest]
+            end,
+    lists:reverse(Result) ++ Rest1;
```

Comment:
We don't discard the existing comment if there is one: beam_validator appears to only acknowledge the last annotation for a variable. What do you think, should we discard the existing comment. I don't imagine it would realistically happen but since the behavior is undocumented it could come back to bite us if we don't discard.

Reply:
Yeah, I think we should merge comments and discard the previous one. As you say, duplicates could be rejected in the future. Also, it makes the analysis of the generated assembly more straightforward. I mean, duplicate comments could be confusing.